tsl/src/compression/algorithms/bool_compress.c

```diff
@@ -0,0 +1,430 @@
+/*
+ * This file and its contents are licensed under the Timescale License.
+ * Please see the included NOTICE for copyright information and
+ * LICENSE-TIMESCALE for a copy of the license.
+ */
+
+#include "bool_compress.h"
+#include "compression/arrow_c_data_interface.h"
+#include "compression/compression.h"
+#include "simple8b_rle.h"
+#include "simple8b_rle_bitmap.h"
+
+typedef struct BoolCompressed
+{
+	CompressedDataHeaderFields;
+	uint8 has_nulls; /* 1 if this has a NULLs bitmap after the values, 0 otherwise */
+	uint8 padding[2];
+	char values[FLEXIBLE_ARRAY_MEMBER];
+} BoolCompressed;
+
+typedef struct BoolDecompressionIterator
+{
+	DecompressionIterator base;
+	Simple8bRleBitmap values;
+	Simple8bRleBitmap nulls;
+	int32 position;
+} BoolDecompressionIterator;
+
+typedef struct BoolCompressor
+{
+	Simple8bRleCompressor values;
+	Simple8bRleCompressor nulls;
+	bool has_nulls;
+	bool last_value;
+} BoolCompressor;
+
+typedef struct ExtendedCompressor
+{
+	Compressor base;
+	BoolCompressor *internal;
+} ExtendedCompressor;
+
+/*
+ * Local helpers
+ */
+static void bool_compressor_append_bool(Compressor *compressor, Datum val);
+
+static void bool_compressor_append_null_value(Compressor *compressor);
+
+static void *bool_compressor_finish_and_reset(Compressor *compressor);
+
+const Compressor bool_compressor_initializer = {
+	.append_val = bool_compressor_append_bool,
+	.append_null = bool_compressor_append_null_value,
+	.finish = bool_compressor_finish_and_reset,
+};
+
+static BoolCompressed *bool_compressed_from_parts(Simple8bRleSerialized *values,
+												  Simple8bRleSerialized *nulls);
+
+static void decompression_iterator_init(BoolDecompressionIterator *iter, void *compressed,
+										Oid element_type, bool forward);
+
+/*
+ * Compressor framework functions and definitions for the bool_compress algorithm.
+ */
+
+extern BoolCompressor *
+bool_compressor_alloc(void)
+{
+	BoolCompressor *compressor = palloc0(sizeof(*compressor));
+	simple8brle_compressor_init(&compressor->values);
+	simple8brle_compressor_init(&compressor->nulls);
+	return compressor;
+}
+
+extern void
+bool_compressor_append_null(BoolCompressor *compressor)
+{
+	compressor->has_nulls = true;
+	simple8brle_compressor_append(&compressor->values, compressor->last_value);
```

Comment:
```suggestion
    /*
     * We use parallel bitmaps of same size for validity and values, to support
     * zero-copy decompression into ArrowArray. When an element is null,
     * the particular value that goes into the values bitmap doesn't matter, so
     * we add the last seen value, not to break the RLE sequences.
     */
	simple8brle_compressor_append(&compressor->values, compressor->last_value);
```
I'd expand a little on the motivation, something like this.

Reply:
Done